File: ui/lib/plan.ts
import type { Bed, Plan, PlanEdit, Stage } from "./types";

// Pure plan helpers — shared by the server (loop, seed) and the client (UI).

export const WORD_BUDGET = 600;

export const STAGES: Stage[] = [
  "empty",
  "seed",
  "seedling",
  "vegetative",
  "flowering",
  "fruiting",
  "harvest",
  "dormant",
];

/** Frost-tender crops the gardener protects when the forecast drops near freezing. */
export const FROST_TENDER = ["pepper", "tomato", "basil", "squash", "cucumber", "bean", "eggplant", "zucchini"];

export function isFrostTender(crop: string): boolean {
  const c = crop.toLowerCase();
  return FROST_TENDER.some((t) => c.includes(t));
}

/** The leanness metric: words across every string in the plan. */
export function planWords(plan: Plan): number {
  let n = 0;
  const walk = (v: unknown) => {
    if (typeof v === "string") n += v.split(/\s+/).filter(Boolean).length;
    else if (Array.isArray(v)) v.forEach(walk);
    else if (v && typeof v === "object") Object.values(v).forEach(walk);
  };
  walk(plan);
  return n;
}

export function emptyPlan(): Plan {
  return {
    beds: [
      { id: "bed_1", crop: "", stage: "empty", sun: "high", water: "med", note: "Full sun, loamy" },
      { id: "bed_2", crop: "", stage: "empty", sun: "med", water: "low", note: "Heavy clay, low corner" },
      { id: "bed_3", crop: "", stage: "empty", sun: "high", water: "med", note: "Trellis on north edge" },
      { id: "bed_4", crop: "", stage: "empty", sun: "high", water: "med", note: "Sandy, drains fast" },
      { id: "bed_5", crop: "", stage: "empty", sun: "high", water: "high", note: "Warmest bed, south wall" },
      { id: "bed_6", crop: "", stage: "empty", sun: "med", water: "med", note: "Afternoon shade" },
    ],
    tasks: [],
    threats: [],
    season_notes: [],
    learnings: [],
  };
}

type Collection = "beds" | "tasks" | "threats";
const COLLECTIONS: Collection[] = ["beds", "tasks", "threats"];
type ListKey = "season_notes" | "learnings";

/**
 * Applies one edit to a plan (returns a new plan, never mutates).
 * Targets: "beds/<id>", "tasks/<id>", "threats/<id>", "season_notes", "learnings".
 */
export function applyEdit(plan: Plan, edit: Pick<PlanEdit, "op" | "target" | "before" | "after">): Plan {
  const next: Plan = structuredClone(plan);
  if (edit.op === "KEEP") return next;
  const [coll, id] = edit.target.split("/") as [string, string | undefined];

  if ((COLLECTIONS as string[]).includes(coll) && id) {
    const list = next[coll as Collection] as { id: string }[];
    const i = list.findIndex((x) => x.id === id);
    if (edit.op === "ADD") {
      if (i >= 0) list[i] = { ...(edit.after as object), id } as { id: string };
      else list.push({ ...(edit.after as object), id } as { id: string });
    } else if (edit.op === "UPDATE" && i >= 0) {
      list[i] = { ...list[i], ...(edit.after as object) };
    } else if (edit.op === "RETIRE" && i >= 0) {
      list.splice(i, 1);
    }
    return next;
  }

  if (coll === "season_notes" || coll === "learnings") {
    const list = next[coll as ListKey];
    const at = typeof edit.before === "string" ? list.indexOf(edit.before) : -1;
    if (edit.op === "ADD" && typeof edit.after === "string") list.push(edit.after);
    else if (edit.op === "UPDATE" && at >= 0 && typeof edit.after === "string") list[at] = edit.after;
    else if (edit.op === "RETIRE" && at >= 0) list.splice(at, 1);
  }
  return next;
}

export function bedNumber(id: string): string {
  return id.replace("bed_", "");
}

export function bedLabel(bed: Bed): string {
  return bed.crop ? capitalize(bed.crop) : "Resting";
}

export function capitalize(s: string): string {
  return s ? s[0].toUpperCase() + s.slice(1) : s;
}

export function isoDay(d: Date): string {
  return d.toISOString().slice(0, 10);
}

export function addDays(d: Date, n: number): Date {
  const x = new Date(d);
  x.setUTCDate(x.getUTCDate() + n);
  return x;
}

/** Next occurrence of weekday (0=Sun … 4=Thu), at least `minAhead` days out. */
export function nextWeekday(from: Date, weekday: number, minAhead = 1): Date {
  let d = addDays(from, minAhead);
  while (d.getUTCDay() !== weekday) d = addDays(d, 1);
  return d;
}

export const WEEKDAY = ["Sun", "Mon", "Tue", "Wed", "Thu", "Fri", "Sat"];

File: loop/plan.py
"""Pure plan helpers — a line-for-line port of ui/lib/plan.ts so both sides agree."""

from __future__ import annotations

import copy
from datetime import date, datetime, timedelta, timezone
from typing import Any

WORD_BUDGET = 600
STAGES = ("empty", "seed", "seedling", "vegetative", "flowering", "fruiting", "harvest", "dormant")
LEVELS = ("low", "med", "high")
FROST_TENDER = ("pepper", "tomato", "basil", "squash", "cucumber", "bean", "eggplant", "zucchini")
WEEKDAY = ("Mon", "Tue", "Wed", "Thu", "Fri", "Sat", "Sun")  # date.weekday() order
COLLECTIONS = ("beds", "tasks", "threats")
LISTS = ("season_notes", "learnings")


def is_frost_tender(crop: str) -> bool:
    c = crop.lower()
    return any(t in c for t in FROST_TENDER)


def plan_words(plan: Any) -> int:
    """The leanness metric: words across every string in the plan."""
    if isinstance(plan, str):
        return len(plan.split())
    if isinstance(plan, list):
        return sum(plan_words(v) for v in plan)
    if isinstance(plan, dict):
        return sum(plan_words(v) for v in plan.values())
    return 0


def apply_edit(plan: dict, edit: dict) -> dict:
    """Applies one edit (never mutates). Targets: beds/<id>, tasks/<id>, threats/<id>, season_notes, learnings."""
    nxt = copy.deepcopy(plan)
    op = edit["op"]
    if op == "KEEP":
        return nxt
    coll, _, ident = edit["target"].partition("/")

    if coll in COLLECTIONS and ident:
        items = nxt[coll]
        i = next((k for k, x in enumerate(items) if x.get("id") == ident), -1)
        after = edit.get("after") or {}
        if op == "ADD":
            row = {**after, "id": ident}
            if i >= 0:
                items[i] = row
            else:
                items.append(row)
        elif op == "UPDATE" and i >= 0:
            items[i] = {**items[i], **after}
        elif op == "RETIRE" and i >= 0:
            items.pop(i)
        return nxt

    if coll in LISTS:
        items = nxt[coll]
        before, after = edit.get("before"), edit.get("after")
        at = items.index(before) if isinstance(before, str) and before in items else -1
        if op == "ADD" and isinstance(after, str):
            items.append(after)
        elif op == "UPDATE" and at >= 0 and isinstance(after, str):
            items[at] = after
        elif op == "RETIRE" and at >= 0:
            items.pop(at)
    return nxt


def find(plan: dict, target: str) -> Any:
    coll, _, ident = target.partition("/")
    if coll in COLLECTIONS and ident:
        return next((x for x in plan.get(coll, []) if x.get("id") == ident), None)
    return None


def next_weekday(start: date, weekday: int, min_ahead: int = 1) -> date:
    """Next occurrence of weekday (Mon=0 … Thu=3), at least ``min_ahead`` days out."""
    d = start + timedelta(days=min_ahead)
    while d.weekday() != weekday:
        d += timedelta(days=1)
    return d


def iso(dt: datetime) -> str:
    """UTC timestamp in the same format as JavaScript's toISOString() — rows compare as strings."""
    return dt.astimezone(timezone.utc).isoformat(timespec="milliseconds").replace("+00:00", "Z")


def empty_plan() -> dict:
    beds = [
        ("bed_1", "high", "med", "Full sun, loamy"),
        ("bed_2", "med", "low", "Heavy clay, low corner"),
        ("bed_3", "high", "med", "Trellis on north edge"),
        ("bed_4", "high", "med", "Sandy, drains fast"),
        ("bed_5", "high", "high", "Warmest bed, south wall"),
        ("bed_6", "med", "med", "Afternoon shade"),
    ]
    return {
        "beds": [{"id": i, "crop": "", "stage": "empty", "sun": s, "water": w, "note": n} for i, s, w, n in beds],
        "tasks": [],
        "threats": [],
        "season_notes": [],
        "learnings": [],
    }
